### PR TITLE
Anon union experiment

### PR DESCRIPTION
**Pull request by realvictorprm**
Currently implements basic simplification of `type Foo = (A_1 | A_2 | ... | A_n)` to `Choice<A_1, A_2, ..., A_n>`.

This is somehow an experiment of what can be done in the direction of non-tagged unions with F#.

### CODE REVIEW

**Comment by Happypig375** [comment]
Yes! The feature I am looking forward to!

**Comment by realvictorprm** [comment]
Stuff is looking good so far, code changes for the following were minimal.
image
These Basic type definitions compiles down to Choice already. Minimal simplification and lexicographic ordering is done too but especially the latter needs better code. Calling `ToString` on `TType` feels a bit akward 😅 .

Besides further support for that syntax is needed but I'm not sure what stuff is possible and good.

My preference would be to be able to do matching on a choice/anon union like this later:
```fsharp
let someFunc (inp: (float | int)) =
   match inp with
   | float a -> printfn "%f" a
   | int a -> printfn "%d" a
```

**Comment by realvictorprm** [comment]
Hm I need some help on getting `type Foo = (unit)`  compile again. Unfortunately it's giving the new rule I added higher priority.

**Comment by abelbraaksma** [comment]
You may want to ping @dsyme or @cartermp, they may know more if such details.

I like where this is going, btw, very interesting. But it's there an RFC already, or lang suggestion, or are you mainly experimenting / trying a poc?

**Comment by realvictorprm** [comment]
There is no RFC for this yet. Only found a suggestion which with the current state is already implemented but for different reasons. I want to try out whether we can get this similar to how Scala is doing it [here](https://d-d.me/talks/scalaworld2015/#/11).

That stuff is a bit more powerful and allows reusing existing union cases kinda type algebra?
E.g. in the end `(int | float | (int64 | int16)` would be equal to `(int | float | int64 | int16)` if I remember the details correctly.

However although I managed to find out that I've put my parser changes to the wrong location & moving them to the correct location I still face a parser issue.
The rule which matches the untagged unions is preferred for matching over the rule matching single type definitions for type abbreviations.
In short `type A = (unit)` gives a parser error with point to that `|` was expected 🤦‍♂️  . I did not find any solution to this yet although I spend a couple of hours on this... 

Any ideas on how I can give the rule for matching `LBRACE typ rbrace` in `atomType` higher precedence?

**Comment by realvictorprm** [comment]
Forgot to cc @dsyme @cartermp. Maybe one of you can give me a hint so I can manage to fix that. That would make me quite happy 😸

**Comment by realvictorprm** [comment]
Ok nevermind. Found a solution for now and will push that later today

**Comment by realvictorprm** [comment]
Ok the last commit should be the fix for the parsing issues.
I would like to discuss possible pattern matching syntax ideas with the F# community.

From your interest @Happypig375 , @abelbraaksma and @drhumlen I want to ask specifically you all for a discussion in the F# slack. I'm not sure about your usernames there so you can find me there with the same name as here on github (realvictorprm) :)

Let the discussion round begin 😁

**Comment by abelbraaksma** [comment]
> I want to ask specifically you all for a discussion in the F# slack

@realvictorprm, I totally missed your message here. I only recently joined Slack, so it's likely you didn't find me there. My handle is almost the same as here: Abel Braaksma (a space in-between). You can either ping me directly, or tag me in one of the #general, #code etc channels.

**Comment by Swoorup** [comment]
Echoing myself here from https://github.com/fsharp/fslang-suggestions/issues/538#issuecomment-635887537.

Basically I am particularly interested in erased union, which only allows distinct types. We discussed a bit on slack too. Duplicates would be a compiler error. Everything here compiles to `System.Object`. 

```fsharp

type Msg = (int | int) // not allowed

type Root = (int | string | decimal) 
type Child = (int | decimal) 
let child: Child = 4
let child2: int = 0
let root: Root = child // implicitly promoted at compilation stage
let root2: Root = child2 // also fine
let child3: Child = Root // error, use pattern matching.

type Protocol = (AddSubscription | RemoveSubscription | VerifyIntegrity | _ )
let a: AddSubscription = { Subscriber = current, job = 0 }
let msg: Protocol = a

let b: Object = int
let msg2: Protocol = b // when converting back compiler forces pattern matching. 
```

This is sort of like enums but on the type level.

**Comment by NatElkins** [comment]
Agree that it should all compile down to `obj` in the end.  Would also be nice if `[<Struct>]` could be added to the erased union and in that case it would compile down to `System.ValueType` and only types that inherited from from `System.ValueType` would be permitted as one of the accepted types.

**Comment by cartermp** [comment]
Closing this out as it's being done as a proper RFC implementation
